rust/src/types/u256_mod.rs

```diff
@@ -0,0 +1,516 @@
+use std::{
+    cmp::Ordering,
+    fmt::Debug,
+    mem,
+    ops::{
+        Add, AddAssign, BitAnd, BitOr, BitXor, Deref, DerefMut, Div, DivAssign, Mul, MulAssign,
+        Not, Rem, RemAssign, Shl, Shr, Sub, SubAssign,
+    },
+};
+
+use bnum::{
+    types::{I256, U256, U512},
+    BInt, BUint,
+};
+use evmc_vm::{Address, Uint256};
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Copy)]
+#[repr(transparent)]
+pub struct u256(Uint256);
+
+impl Deref for u256 {
+    type Target = [u8; 32];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0.bytes
+    }
+}
+
+impl DerefMut for u256 {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0.bytes
+    }
+}
+
+impl Debug for u256 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("0x")?;
+        for (i, byte) in (*self).into_iter().enumerate() {
+            f.write_fmt(format_args!("{:02x}", byte))?;
+            if i % 8 == 7 && i < 31 {
+                f.write_str("_")?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl u256 {
+    pub const ZERO: Self = Self(Uint256 { bytes: [0; 32] });
+    pub const MAX: Self = Self(Uint256 { bytes: [0xff; 32] });
+}
+
+impl From<Uint256> for u256 {
+    fn from(value: Uint256) -> Self {
+        Self(value)
+    }
+}
+
+impl From<u256> for Uint256 {
+    fn from(value: u256) -> Self {
+        value.0
+    }
+}
+
+impl From<U256> for u256 {
+    fn from(value: U256) -> Self {
+        // TODO bnum has to_be_bytes with feature nightly
+        let be_value = value.to_be();
+        let bytes: [u8; 32] = unsafe {
+            // U256 = BUint<4>
+            // BUint<4> is transparent wrapper around [u64; 4]
+            mem::transmute(be_value)
+        };
+        bytes.into()
+    }
+}
+
+impl From<u256> for U256 {
+    fn from(value: u256) -> Self {
+        BUint::from_be_slice(value.deref()).unwrap()
+    }
+}
+
+impl From<I256> for u256 {
+    fn from(value: I256) -> Self {
+        // TODO bnum has to_be_bytes with feature nightly
+        let be_value = value.to_be();
+        let bytes: [u8; 32] = unsafe {
+            // I256 = is transparent wrapper around U256
+            // U256 = BInt<4>
+            // BUint<4> is transparent wrapper around [u64; 4]
+            mem::transmute(be_value)
+        };
+        bytes.into()
+    }
+}
+
+impl From<u256> for I256 {
+    fn from(value: u256) -> Self {
+        BInt::from_be_slice(value.deref()).unwrap()
+    }
+}
+
+impl From<U512> for u256 {
```

Comment:
Not sure if we need this at all. I think we can drop 512 and 128 for the time being.

Reply:
this is for addmod and mulmod where the intermediate value (sum or product) may not fit into 256 bits